common/src/chain/transaction/signature.rs

```diff
@@ -0,0 +1,254 @@
+            ],
+            vec![TxOutput::new(Amount::from_atoms(100), dest.clone())],
+            0,
+        )?;
+
+        let signature = StandardInputSignature::produce_signature_for_input(
+            &private_key,
+            sighash_type,
+            dest.clone(),
+            &tx,
+            0,
+        )
+        .unwrap();
+
+        tx.update_witness(0, InputWitness::Standard(signature.clone())).unwrap();
+        Ok(tx)
+    }
+
+    fn sign_tx(
+        tx: &mut Transaction,
+        private_key: &PrivateKey,
+        sighash_type: SigHashType,
+        dest: Destination,
+    ) {
+        for i in 0..tx.get_inputs().len() {
+            let input_sign = StandardInputSignature::produce_signature_for_input(
+                private_key,
+                sighash_type,
+                dest.clone(),
+                &tx,
+                0,
+            )
+            .unwrap();
+            tx.update_witness(i, InputWitness::Standard(input_sign)).unwrap();
+        }
+    }
+
+    #[test]
+    #[allow(clippy::eq_op)]
+    fn sign_and_verify() {
+        let (private_key, public_key) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+        let outpoint_dest = Destination::PublicKey(public_key);
+        let sighash_type = SigHashType::try_from(SigHashType::ALL).unwrap();
+        let mut tx = generate_tx(private_key.clone(), sighash_type, outpoint_dest.clone()).unwrap();
+        sign_tx(&mut tx, &private_key, sighash_type, outpoint_dest.clone());
+        assert!(verify_signature(&outpoint_dest, &tx, 0).is_ok());
+        assert!(verify_signature(&outpoint_dest, &tx, 1).is_ok());
+        // ALL
+        let sighash_type = SigHashType::try_from(SigHashType::ALL).unwrap();
+        let tx = generate_tx(private_key.clone(), sighash_type, outpoint_dest.clone()).unwrap();
+        assert!(verify_signature(&outpoint_dest, &tx, 0).is_ok());
+        assert!(verify_signature(&outpoint_dest, &tx, 1).is_err());
+        // NONE
+        let sighash_type = SigHashType::try_from(SigHashType::NONE).unwrap();
+        let tx = generate_tx(private_key.clone(), sighash_type, outpoint_dest.clone()).unwrap();
+        assert!(verify_signature(&outpoint_dest, &tx, 0).is_ok());
+        assert!(verify_signature(&outpoint_dest, &tx, 1).is_err());
+        // SINGLE
+        let sighash_type = SigHashType::try_from(SigHashType::SINGLE).unwrap();
+        let tx = generate_tx(private_key.clone(), sighash_type, outpoint_dest.clone()).unwrap();
+        assert!(verify_signature(&outpoint_dest, &tx, 0).is_ok());
+        assert!(verify_signature(&outpoint_dest, &tx, 1).is_err());
+        // ANYONECANPAY
+        let sighash_type =
+            SigHashType::try_from(SigHashType::ALL | SigHashType::ANYONECANPAY).unwrap();
+        let tx = generate_tx(private_key.clone(), sighash_type, outpoint_dest.clone()).unwrap();
+        assert!(verify_signature(&outpoint_dest, &tx, 0).is_ok());
+        assert!(verify_signature(&outpoint_dest, &tx, 1).is_err());
```

Comment:
There should also be tests that mutate:
1. The signature
2. The data behind the signature, one by one

So you try to manipulate the signature by botching a byte (or a bit) in it. Then you ensure that verification will fail. Be careful in such tests as they can become flaky if done incorrectly. For example, don't set a single byte to a specific value you give, like `'x'` always, because then if the byte value is `'x'`, the test will basically change nothing. A good solution is using a `wrapping_add()` on a random byte` in the signature.

Then, you have multiple components that we signed. We want to test that manipulating them isn't really possible. For every type. For example, you change the amounts, number of outputs, etc. See what signature hash is hashing, and try to change everything that a specific signature tries to seal, and modify that.

Reply:
Look at the code please, is it enough?

Reply:
Things are getting better, but, not everything is covered yet.

For every sighash type, please list in a comment the contents, and then modify them one by one. I see only SighashType::ALL. This certainly isn't enough.